src/core/api/scan.go

```diff
@@ -0,0 +1,199 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/goharbor/harbor/src/pkg/scan/report"
+
+	"github.com/goharbor/harbor/src/common/models"
+	"github.com/goharbor/harbor/src/common/rbac"
+	"github.com/goharbor/harbor/src/common/utils"
+	coreutils "github.com/goharbor/harbor/src/core/utils"
+	"github.com/goharbor/harbor/src/jobservice/logger"
+	"github.com/goharbor/harbor/src/pkg/scan/api/scan"
+	v1 "github.com/goharbor/harbor/src/pkg/scan/rest/v1"
+	"github.com/pkg/errors"
+)
+
+var digestFunc digestGetter = getDigest
+
+// ScanAPI handles the scan related actions
+type ScanAPI struct {
+	BaseController
+
+	// Target artifact
+	artifact *v1.Artifact
+	// Project reference
+	pro *models.Project
+}
+
+// Prepare sth. for the subsequent actions
+func (sa *ScanAPI) Prepare() {
+	// Call super prepare method
+	sa.BaseController.Prepare()
+
+	// Parse parameters
+	repoName := sa.GetString(":splat")
+	tag := sa.GetString(":tag")
+	projectName, _ := utils.ParseRepository(repoName)
+
+	pro, err := sa.ProjectMgr.Get(projectName)
+	if err != nil {
+		sa.SendInternalServerError(errors.Wrap(err, "scan API: prepare"))
+		return
+	}
+	if pro == nil {
+		sa.SendNotFoundError(errors.Errorf("project %s not found", projectName))
+		return
+	}
+	sa.pro = pro
+
+	// Check authentication
+	if !sa.SecurityCtx.IsAuthenticated() {
+		sa.SendUnAuthorizedError(errors.New("Unauthorized"))
+		return
+	}
+
+	// Assemble artifact object
+	digest, err := digestFunc(repoName, tag, sa.SecurityCtx.GetUsername())
+	if err != nil {
+		sa.SendInternalServerError(errors.Wrap(err, "scan API: prepare"))
+		return
+	}
+
+	sa.artifact = &v1.Artifact{
+		NamespaceID: pro.ProjectID,
+		Repository:  repoName,
+		Tag:         tag,
+		Digest:      digest,
+		MimeType:    v1.MimeTypeDockerArtifact,
+	}
+
+	logger.Debugf("Scan API receives artifact: %#v", sa.artifact)
+}
+
+// Scan artifact
+func (sa *ScanAPI) Scan() {
+	// Check access permissions
+	resource := rbac.NewProjectNamespace(sa.pro.ProjectID).Resource(rbac.ResourceScan)
+	if !sa.SecurityCtx.Can(rbac.ActionCreate, resource) {
+		sa.SendForbiddenError(errors.New(sa.SecurityCtx.GetUsername()))
+		return
+	}
```

Comment:
There is a method RequireProjectAccess in BaseController, that can do permission checking simpler.